Implement `ssh` task for opening ssh connection and going to a release path

pluseg opened the pull request:
| Q             | A
| ------------- | ---
| Bug fix?      | No
| New feature?  | Yes
| BC breaks?    | No
| Deprecations? | No
| Fixed tickets | #1013 

After review and approval I'll add a PR into docs repo.

elfet commented:
Cool. But better use posix exec.

elfet commented on recipe/common.php:
```diff
@@ -128,6 +130,22 @@
 });
 
+desc('Connect to ssh and `cd` to release path');
+task('ssh', function () {
```
Better to create it as ssh console command, not task.

pluseg replied:
@elfet could you please explain the better side of this decision? I'm pretty new in the project so I can miss something important.

elfet commented on src/Server/SSHPipeInterface.php:
```diff
@@ -0,0 +1,19 @@
+
+namespace Deployer\Server;
+
+interface SSHPipeInterface
```
This it really strange interface.

pluseg replied:
What's strange for you? I think that there is a chance in the future that we need to add new type of `Server`, so we need to define which of them supports `ssh` task and which one does not. If we have an interface then we can be sure that a class implemented one has concrete method like `createSshPipe()`. Or do you think that checking `getenv('ssh_type')` is enough?

pluseg replied:
@elfet I don't understand what I should do here. My position is written down above. I think that name of the interface might be better but I have no idea.

pluseg commented:
@elfet 

> But better use posix exec.

Do you mean [pcntl_exec](http://php.net/manual/en/function.pcntl-exec.php)? Yeah, it's the best solution but it requires `pcntl` extension. Also PHP.net says "Currently, this module will not function on non-Unix platforms (Windows)". If you mean simple `exec()` function then it doesn't work for me, terminal just freezing. Is there another solution that I missed?

elfet commented:
Use pcntl_exec if possible otherwise proc_open.

pluseg commented:
Mini-changelog:
1. Added pcntl support for running `ssh` process separately.
2. pcntl_exec gets different arguments and escape them rather than proc_open. So it leads to light refactoring of building ssh command line options.
3. Added `ssh` console command for ssh connections
4. The item above leads to moving getting stage, tag and revision options to TaskCommand. I guess that it's more suitable place.

@elfet do i need to reflect all these items in CHANGELOG.md or not? I added only third item as the most valuable.

pluseg commented on recipe/common.php:
```diff
@@ -113,14 +115,6 @@
-option('revision', null, InputOption::VALUE_OPTIONAL, 'Revision to deploy');
-option('branch', null, InputOption::VALUE_OPTIONAL, 'Branch to deploy');
-
-/**
```
I moved this section into TaskCommand. I have no idea why they are called here. @elfet do you have any thoughts here?

pluseg commented on src/Server/Configuration.php:
```diff
@@ -340,6 +340,16 @@ public function setUser($user)
+     */
+    public function getUserAndHost()
+    {
+        return (!empty($this->getUser()) ? $this->getUser() . '@' : '') . $this->getHost();
```
It's used in several places so I create separate method.

pluseg commented on src/Server/Remote/NativeSsh.php:
```diff
@@ -141,23 +99,25 @@ public function scpCopy($target, $target2)
         if ($serverConfig->getPrivateKey()) {
-            $scpOptions[] = '-i ' . escapeshellarg($serverConfig->getPrivateKey());
+            $scpOptions = array_merge_recursive($scpOptions, ['-i' => $serverConfig->getPrivateKey()]);
+        } elseif ($serverConfig->getPemFile()) {
```
I've noticed that pem file support for ssh connection was added recently but it was missed for scp.

pluseg commented on src/Server/Remote/NativeSsh.php:
```diff
@@ -230,12 +206,55 @@ protected function getConnectionHash()
+                'ControlMaster=auto',
+                'ControlPersist=5',
+                'ControlPath=' . $this->getConnectionHash(),
+            ],
```
I changed structure of options due to new way of building command line arguments.

pluseg commented on src/Server/Remote/NativeSsh.php:
```diff
@@ -244,32 +263,118 @@ protected function getMultiplexingSshOptions()
                 \Deployer\writeln('  SSH multiplexing initialization');
             }
-            exec('ssh ' . implode(' ', $sshOptions) . ' ' . escapeshellarg($username . $hostname) . "  'echo \"SSH multiplexing initialization\"' ");
+            exec('ssh ' . $this->buildCommandLineOptions($sshOptions, true) . "  'echo \"SSH multiplexing initialization\"' ");
```
Actually I didn't test it after options refactoring because I'm not sure how to do that correctly.

pluseg commented on src/Server/Remote/NativeSsh.php:
```diff
@@ -244,32 +263,118 @@ protected function getMultiplexingSshOptions()
+         * [HACK] PCNTL works only on Unix and `pcntl_exec` requires full path of running command.
+         * So it defines path of `ssh` using `which` utility.
+         */
+        $sshPath = trim(shell_exec('which ssh'));
```
I have no idea how to do it more cross-platform.

ossinkine commented on src/Server/Remote/NativeSsh.php:
```diff
@@ -230,12 +206,55 @@ protected function getConnectionHash()
+     *
+     * @return array
+     */
+    private function getSshConnectionOptions()
```
I think this these options should be reusable. Maybe make this method public?
For example, I use rsync in my docker.php and it also needs these options. How should I pass them into it?

pluseg replied:
@ossinkine through configuration in deploy.php.

pluseg replied:
@ossinkine In fact we need to create separate AbstractOptionsBuilder and SshOptionsBuilder class for working with all that stuff. But the PR is already too big so I'm going to do it next time.

elfet commented:
A lot of refactoring. Will try t review it all. Or maybe will cherry pick some of your code to new v5 branch.
Great work, by the way.

pluseg commented:
This refactoring isn't final. There is still code duplication and high complexity. But now it is prepared for further improvements. E.g we can move command line options to separate class which requires only $this->configuration and can build options for different targets (scp, ssh, etc).

elfet commented:
@pluseg okay. I can finish your pr by mysql. 🙂

antonmedv commented:
Done in #1092
